src/commands/serve.rs

```diff
@@ -94,38 +117,127 @@ impl ServeCommand {
+                    .collect::<Vec<_>>();
+                let (backends, registry) = wasmtime_wasi_nn::preload(&graphs)?;
+                host.nn.replace(WasiNnCtx::new(backends, registry));
+            }
+        }
+
+        let mut store = Store::new(engine, host);
+
+        if self.run.common.wasm.timeout.is_some() {
+            store.set_epoch_deadline(1);
+        }
+
+        store.data_mut().limits = self.run.store_limits();
+        store.limiter(|t| &mut t.limits);
+
+        // If fuel has been configured, we want to add the configured
+        // fuel amount to this store.
+        if let Some(fuel) = self.run.common.wasm.fuel {
+            store.add_fuel(fuel)?;
+        }
+
+        Ok(store)
+    }
+
     fn add_to_linker(&self, linker: &mut Linker<Host>) -> Result<()> {
+        // wasi-http and the component model are implicitly enabled for `wasmtime serve`, so we
+        // don't test for `self.run.common.wasi.common` or `self.run.common.wasi.http` in this
+        // function.
+
         wasmtime_wasi_http::proxy::add_to_linker(linker)?;
+
+        if self.run.common.wasi.nn == Some(true) {
+            #[cfg(feature = "wasi-nn")]
+            {
+                wasmtime_wasi_nn::wit::ML::add_to_linker(linker, |host| host.nn.as_mut().unwrap())?;
+            }
+        }
+
         Ok(())
     }
 
-    async fn serve(&mut self) -> Result<()> {
+    async fn serve(mut self) -> Result<()> {
         use hyper::server::conn::http1;
 
-        let mut config = self.common.config(None)?;
+        self.run.common.init_logging();
+
+        let mut config = self.run.common.config(None)?;
         config.wasm_component_model(true);
         config.async_support(true);
 
-        let engine = Arc::new(Engine::new(&config)?);
+        if self.run.common.wasm.timeout.is_some() {
+            config.epoch_interruption(true);
+        }
+
+        match self.run.profile {
+            Some(Profile::Native(s)) => {
+                config.profiler(s);
+            }
+
+            // We bail early in `execute` if the guest profiler is configured.
+            Some(Profile::Guest { .. }) => unreachable!(),
+
+            None => {}
+        }
+
+        let engine = Engine::new(&config)?;
         let mut linker = Linker::new(&engine);
 
         self.add_to_linker(&mut linker)?;
 
         let component = Component::from_file(&engine, &self.component)?;
 
-        let instance = Arc::new(linker.instantiate_pre(&component)?);
+        let instance = linker.instantiate_pre(&component)?;
 
-        let listener = tokio::net::TcpListener::bind(self.addr()).await?;
+        let listener = tokio::net::TcpListener::bind(self.addr).await?;
+
+        let _epoch_thread = if let Some(timeout) = self.run.common.wasm.timeout {
+            let engine = engine.clone();
+            Some(preview2::spawn(async move {
+                tokio::time::sleep(timeout).await;
+                engine.increment_epoch();
+            }))
```

Comment:
Viceroy spawns a thread for this, is it reasonable to use a tokio task instead?

Reply:
No I think this is best done as a thread because if you've got N cores each running wasm within Tokio then because nothing is preemptive a separate Tokio task wouldn't ever actually run to increment the epoch, so OS scheduling is required to guarantee epochs get incremented and Tokio tasks yield